### PR TITLE
Improvements of interactions with media in reply messages

### PR DESCRIPTION
**Pull request by grimhilt**
Fixes https://github.com/vector-im/element-web/issues/20150
Improves https://github.com/vector-im/element-web/issues/22911

Improvements of media in reply messages :
* Voice messages
    * Allows clicking on the play pause button and the seek bar of a voice message without being jumped to the original message
    * Clicking everywhere else will still result in a jump
    * Playing a voice message in a reply will not start a playback queue
* Images
    * Clicking on the preview of an image inside a reply will open it
    * Clicking everywhere else will still jump you to the original message
* Media files (including video and audio files)
    * Clicking on the title or button of a file will no longer download it
    * But it will jump you to the origin message

Signed-off-by: [grimhilt@users.noreply.github.com](mailto:grimhilt@users.noreply.github.com)

---
Here's what your changelog entry will look like:

## ✨ Features
 * Improvements of interactions with media in reply messages ([\#9069](https://github.com/matrix-org/matrix-react-sdk/pull/9069)). Fixes vector-im/element-web#20150. Contributed by @grimhilt.

### CODE REVIEW

File: src/components/views/messages/MFileReplyBody.tsx [hunk 1]
```diff
@@ -0,0 +1,52 @@
+/*
+Copyright 2020-2021 Tulir Asokan <tulir@maunium.net>
```
Comment (turt2live) [inline src/components/views/messages/MFileReplyBody.tsx:2]: This is the wrong copyright

File: src/components/views/messages/MVoiceMessageReplyBody.tsx [hunk 2]
```diff
@@ -0,0 +1,55 @@
+/*
+Copyright 2021 The Matrix.org Foundation C.I.C.
```
Comment (turt2live) [inline src/components/views/messages/MVoiceMessageReplyBody.tsx:2]: This is the wrong copyright

File: src/components/views/messages/MVoiceMessageReplyBody.tsx [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+import React from "react";
+
+import InlineSpinner from '../elements/InlineSpinner';
+import { _t } from "../../../languageHandler";
+import RecordingPlayback from "../audio_messages/RecordingPlayback";
+import MAudioBody from "./MAudioBody";
+import MFileBody from "./MFileBody";
+import MediaProcessingError from "./shared/MediaProcessingError";
+
+export default class MVoiceMessageReplyBody extends MAudioBody {
```
Comment (turt2live) [inline src/components/views/messages/MVoiceMessageReplyBody.tsx:26]: I'm a bit confused as to why we're effectively duplicating a component here?

File: src/components/views/messages/MFileReplyBody.tsx [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+import { _t } from "../../../languageHandler";
+import AccessibleButton from "../elements/AccessibleButton";
+import { presentableTextForFile } from "../../../utils/FileUtils";
+
+export default class MFileReplyBody extends MFileBody {
+    protected onPlaceholderClick = async () => {
+        return;
+    };
+
+    render() {
```
Comment (turt2live) [inline src/components/views/messages/MFileReplyBody.tsx:29]:
```suggestion
    public render() {
```

File: src/audio/Playback.ts [hunk 5]
```diff
@@ -61,6 +61,7 @@ export class Playback extends EventEmitter implements IDestroyable {
     private waveformObservable = new SimpleObservable<number[]>();
     private readonly clock: PlaybackClock;
     private readonly fileSize: number;
+    private isInReply: boolean;
```
Comment (turt2live) [inline src/audio/Playback.ts:64]: This shouldn't be tracked off the playback instance - we should be ensuring we're not queueing the playback at all from elsewhere.

From memory, we track the voice message order by pushing on mount, so we should be able to use a component property.

File: src/components/views/rooms/ReplyTile.tsx [hunk 6]
```diff
@@ -79,14 +79,27 @@ export default class ReplyTile extends React.PureComponent<IProps> {
...
-        // The exception being the link wrapping the reply
-        if (
+
+        // Clicking an image should open it and
+        // should not dispatch the `view_room` action
+        if (clickTarget.tagName.toLowerCase() === "img") {
+            e.preventDefault();
+        } else if (
+            clickTarget.classList.contains("mx_SeekBar") ||
+            clickTarget.getAttribute("data-test-id") === "play-pause-button"
```
Comment (turt2live) [inline src/components/views/rooms/ReplyTile.tsx:89]: err, we shouldn't be relying on test IDs in production code. Can we not check for the voice message container?

File: src/components/views/messages/MFileReplyBody.tsx [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+*/
+
+import React from "react";
+
+import MFileBody from "./MFileBody";
+import { _t } from "../../../languageHandler";
+import AccessibleButton from "../elements/AccessibleButton";
+import { presentableTextForFile } from "../../../utils/FileUtils";
+
+export default class MFileReplyBody extends MFileBody {
```
Comment (turt2live) [inline src/components/views/messages/MFileReplyBody.tsx:24]: it's also not clear what this component's purpose is, and it appears to duplicate the code

**Review by turt2live — CHANGES_REQUESTED** [review]
Thanks for the PR! The feature itself seems to be a good idea. 

Please also take a look at the CI failures and resolve those before re-requesting review :)

File: src/components/views/messages/MFileReplyBody.tsx (hunk 7 above)
Comment (grimhilt) [reply src/components/views/messages/MFileReplyBody.tsx:24]: as i didn't find a way to make the MFileBody component aware that it is in a reply, I use this component to override the onPlaceholderClick function and prevent download on click
i will try another way adding a property "isInReply" in IBodyProps to remove this duplication

File: src/components/views/rooms/ReplyTile.tsx (hunk 6 above)
Comment (grimhilt) [reply src/components/views/rooms/ReplyTile.tsx]: I change the test ID for a class test as I expect to be moved in the timeline when I click a non-interactive element of the voice message (e.g clock)

**Comment by grimhilt** [comment]
Thank you for reviewing @turt2live
I still have one error in the tests but i also have it in the branch develop.

**Comment by turt2live** [comment]
something appears to have gone very wrong with your merge - it's now including commits that are already on develop.

**Review by turt2live — COMMENTED** [review]
clearing app team code review

**Comment by grimhilt** [comment]
indeed, i will try to fix this, I think I did a rebase instead of a merge with develop

**Comment by grimhilt** [comment]
I'm closing this as it is outdated and it will be easier to start over.

**Pull request closed without merging**